### ISSUE TITLE
data-annotation-id not added on initial annotation creation

### ISSUE THREAD

**Issue by ACParadox** [openannotation/annotator#495]
The "data-annotation-id" on the highlight span is added on "updateAnnotation" which seems to trigger when the annotations are loaded from storage or modified, but not when the annotation is first created ("createAnnotation).
So, without reloading the page, the "data-annotation-id" is not immediately available after annotation creation.

**Comment by jhou2** [comment]
For my project, I used jquery attr() to add the data-annotation-id after my server returned it.

**Comment by kgish** [comment]
Yes this is a ugly bug. I need to update status of updates when a new annotation is created, but am unable to access `$('#annotation_id')` yet because the returned `annotation.id` is still undefined.

**Comment by tilgovi** [comment]
Easy fix for the v1.2.x branch if someone wants to make a PR.
In the Store plugin's `updateAnnotation` callback, you can update the highlights.

**Comment by tilgovi** [comment]
It's already updating the data of the spans, which may actually be redundant since the data object stored on the highlight is by reference the same object that the Store updates.

**Comment by kgish** [comment]
This is caused by a bug in `addAnnotation` and not `updateAnnotation`. For those interested, here's a workaround:

``` javascript
// Register an ajax complete handler.
$(document).ajaxComplete(function(event, jqXHR, ajaxOptions) {
    if (ajaxOptions.type === 'POST' && /\/annotations$/.test(ajaxOptions.url)) {
        var annotation = JSON.parse(jqXHR.responseText);
        cached_ann.push(annotation);
        // All highlighted annotations should be enclosed within spans like this:
        //     <span class="annotator-hl" data-annotation-id="25" id="annotation_25"
        //         name="annotation_25">...</span>
        // However, the highlighted annotation just created only looks like this
        //     <span class="annotator-hl">...</span>
        // So we need to add attributes: data-annotation-id, id, and name
        var wrapper = $(sel_wrapper);
        if (wrapper.length === 0) {
            console.error("ajaxComplete() => cannot find " +
                "wrapper $(" + sel_wrapper + ")!");
            return;
        }
        var annotations = wrapper.find($('.annotator-hl'));
        if (annotations.length === 0) {
            console.error("ajaxComplete() => cannot any annotations!");
            return;
        }
        annotations.each(function(index, a){
            if (!$(a).attr('id')) {
                var attr = 'annotation_' + annotation.id;
                ['id', 'name', 'data-annotation-id'].forEach(function(name){
                    $(a).attr(name, attr);
                });
                return false;
            }
        });
        display_annotations();
    }
});
```

**Comment by tilgovi** [comment]
I mean `updateAnnotation` the callback in the `Store` plugin. Not the `updateAnnotation` fuction on `Annotator`. I understand that this is happening when you add an annotation, but the `Store` plugin has its own method `updateAnnotation` that is called after its remote request returns.

https://github.com/openannotation/annotator/blob/v1.2.x/src/plugin/store.coffee#L228

This is a one line PR fix in that function.

**Comment by kgish** [comment]
Sorry, maybe I'm blind or just plain stupid, but I cannot see figure out what that one-liner might be.

**Comment by tilgovi** [comment]
I'm certainly not trying to make anyone feel stupid, just trying to lead and encourage you to possibly figuring it out and encourage involvement!

If you look at the link I left there, you should see a function called `updateAnnotation`. This function is is a callback inside the Store plugin that's invoked after both create and update. It's possibly poorly named. Unlike `annotationUpdated`, which is what gets called when the _interface_ requests that an annotation be updated/edited and therefore stored, this is called _after_ the annotation has been created/updated in order to update the interface with whatever result was returned. For our purposes, we're interested in the fact that the act of creating the annotation has now resulted in the server giving it a unique id.

The very last line of that function is this:

``` javascript
    $(annotation.highlights).data('annotation', annotation)
```

That makes sure that all the highlights have the latest copy of the data after the XHR call returns.\* It is at this point that these highlights could also be updated with a call to `.attr()` to add the annotation's id.

*: I suspect this is superfluous, because the object is updated in place anyway and shared among all the highlights' data.
